hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestViewDistributedFileSystemWithMountLinks.java

```diff
@@ -61,4 +64,55 @@ public void testCreateOnRoot() throws Exception {
   public void testMountLinkWithNonExistentLink() throws Exception {
     testMountLinkWithNonExistentLink(false);
   }
+
+  @Test
+  public void testRenameOnInternalDirWithFallback() throws Exception {
+    Configuration conf = getConf();
+    URI defaultFSURI =
+        URI.create(conf.get(CommonConfigurationKeys.FS_DEFAULT_NAME_KEY));
+    final Path hdfsTargetPath1 = new Path(defaultFSURI + "/HDFSUser");
+    final Path hdfsTargetPath2 = new Path(defaultFSURI + "/NewHDFSUser/next");
+    ViewFsTestSetup.addMountLinksToConf(defaultFSURI.getAuthority(),
+        new String[] {"/HDFSUser", "/NewHDFSUser/next"},
+        new String[] {hdfsTargetPath1.toUri().toString(),
+            hdfsTargetPath2.toUri().toString()}, conf);
+    //Making sure parent dir structure as mount points available in fallback.
+    try (DistributedFileSystem dfs = new DistributedFileSystem()) {
+      dfs.initialize(defaultFSURI, conf);
+      dfs.mkdirs(hdfsTargetPath1);
+      dfs.mkdirs(hdfsTargetPath2);
+    }
+
+    try (FileSystem fs = FileSystem.get(conf)) {
+      Path src = new Path("/newFileOnRoot");
+      Path dst = new Path("/newFileOnRoot1");
+      fs.create(src).close();
+      verifyRename(fs, src, dst);
+
+      src = new Path("/newFileOnRoot1");
+      dst = new Path("/NewHDFSUser/newFileOnRoot");
+      fs.mkdirs(dst.getParent());
+      verifyRename(fs, src, dst);
+
+      src = new Path("/NewHDFSUser/newFileOnRoot");
+      dst = new Path("/NewHDFSUser/newFileOnRoot1");
+      verifyRename(fs, src, dst);
+
+      src = new Path("/NewHDFSUser/newFileOnRoot1");
+      dst = new Path("/newFileOnRoot");
+      verifyRename(fs, src, dst);
+
+      src = new Path("/HDFSUser/newFileOnRoot1");
+      dst = new Path("/HDFSUser/newFileOnRoot");
+      fs.create(src).close();
+      verifyRename(fs, src, dst);
+    }
+  }
+
+  private void verifyRename(FileSystem fs, Path src, Path dst)
+      throws IOException {
+    fs.rename(src, dst);
+    Assert.assertFalse(fs.exists(src));
+    Assert.assertTrue(fs.exists(dst));
+  }
```

Comment:
I tried a couple of cases of rename. Can you give a check once.
``` Test
  public void testRenameOnInternalDirWithFallback() throws Exception {
    Configuration conf = getConf();
    URI defaultFSURI =
        URI.create(conf.get(CommonConfigurationKeys.FS_DEFAULT_NAME_KEY));
    final Path hdfsTargetPath1 = new Path(defaultFSURI + "/HDFSUser");
    final Path hdfsTargetPath2 = new Path(defaultFSURI + "/dstNewHDFSUser"
        + "/next");
    ViewFsTestSetup.addMountLinksToConf(defaultFSURI.getAuthority(),
        new String[] {"/HDFSUser", "/NewHDFSUser/next/next1"},
        new String[] {hdfsTargetPath1.toUri().toString(),
            hdfsTargetPath2.toUri().toString()}, conf);
    //Making sure parent dir structure as mount points available in fallback.
    try (DistributedFileSystem dfs = new DistributedFileSystem()) {
      dfs.initialize(defaultFSURI, conf);
      dfs.mkdirs(hdfsTargetPath1);
      dfs.mkdirs(hdfsTargetPath2);
    }

    try (FileSystem fs = FileSystem.get(conf)) {
      // Case : 1
      Path src = new Path("/newFileOnRoot");
      Path dst = new Path("/NewHDFSUser/next");
      fs.create(src).close();
      verifyRename(fs, src, dst); // Fails. Shouldn't it move to
      // /NewHDFSUser/next/newFileOnRoot ?

       src = new Path("/newFileOnRoot");
       dst = new Path("/NewHDFSUser/next/file");
      verifyRename(fs, src, dst); // Fails. Guess since the parent structure
      // isn't there at fallback?```

I couldn't check more, but
CASE 1: if the destination is a directory, shouldn't it move src inside it?
CASE:2 Seems due to parent structure isn't there in fallback?

I didn't try with ViewFs, maybe something similar there as well

Reply:
@ayushtkn , Thanks a lot for the review!
Good findings. 
Case1: Actually there are two rename behaviors, I think rename2 does expect both(src and dst) to be same type either dir or file. However I modified it to allow internalDir as dst with fallback, rename2 can fail later as usual. But the current rename(src,dst) also works as you verified. Please check the modified code whether it can satisfy you expected conditions.
Case2: Yes, this is know fact that, if fallback does not have structure, rename does not create any dirs and it will fail.
When user does not have any fallback structure, I think they should not expect all of this this should work very well as they might have created arbitrary mount point, but not created mount points with respective to fallback/default cluster structure.
Please note ViewFs rename behave like rename2 so, test assertions are slightly different than rename api from ViewFileSystem.

Reply:
Thanx @umamaheswararao for the update.
Regarding the Case 2: When the same directory structure isn't available in the fallback.
In `ViewFs` I think this was handled and `createParent` was explicitly made `true` always. It would be just for rename this compulsion would be there. 
Considering a mount entry like -- `/mount/sub1/sub2` --> `/nsPath`
if someone calls rename with `dst` as `/mount/sub1/renameFile` will fail, but if he calls create `/mount/sub1/createFile` without `createParent` it would pass and this `create` call will create the internal directory structure as well. So, now again the user calls the same rename command, it would succeed. Same for `mkdir` with `createParent` as `false`
This would be little intermittent behavior for the end user, one API behaving differently.
Secondly creating the same directory structure at `fallback` just for `rename` to work doesn't seems feasible, It would be too many empty directories, increasing the number of inodes at NN. IIRC something like this, to create empty directories for mount entries in case of RBF was discussed for some issue recently, and UBER folks had concerns with inode numbers going high due to empty directories.

I think we should explicitly take care of this in `rename` as well, May be in non-atomic way only? Latter we might find a better way, Maybe adding one more flag to `rename2` and argument to `rename` for `createParent` in a follow up.

Reply:
What's your proposal for rename here? Could you elaborate a bit?
I am thinking this way: If API nature is to create parent dirs, it will create. If the API nature is not create parents, I am ok to fail and let the users know the fact that it's redirected to fallback and they don't have structure?
In your above example, you made additional call to createFile, so that next rename succeeded. 
I see in ViewFS.java create, mkdir we made createParent by default true. I am ok to take the user passed flag itself. That way we can be consistent on API perspective. Also if I am correct, this problem only exist with ViewFs.java. In ViewFileSystem.java as we will compliant to API as mkdirs by default create parent. This way, users expected things only we do. What do you? If someone really wants to createParents bydefault in ViewFs.java, that discussion can be taken into separate JIRA for create,mkdir behavior. To summarize: in ViewFs.java, create, mkdir creates the parent in fallback if they don't exist (irrespective of createParent flag. ) We did this because, for the users perspective, we provide transparency and exist of parent dir returns true. However that fact is that, children actually goes to fallback. When users create mount points with respective to fallback cluster, then only this will be true. Otherwise it can be any random mount point without any relation to fallback. When API does not say createParent, then we will keep that semantics to fallback as well. Whats your thoughts?